apis/configwatch.go

```diff
@@ -181,3 +205,101 @@ func MonitorTLS(start time.Time, expiry time.Time, conf *config.Config) {
 		}
 	}
 }
+
+func (s *Server) getClientValidator(helloInfo *tls.ClientHelloInfo) func([][]byte, [][]*x509.Certificate) error {
+
+	log.Debug().Msgf("Inside get client validator - %v", helloInfo.Conn.RemoteAddr())
+	return func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// Verifying client certs with root ca
+		opts := x509.VerifyOptions{
+			Roots:         s.CaCertPool,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		}
+		_, err := verifiedChains[0][0].Verify(opts)
+		if err != nil {
+			log.Error().Err(err).Msgf("certs verification failed")
+			return err
+		}
+
+		log.Debug().Msgf("validating with SAN match rules - %s", s.SANMatchRules)
+		if len(s.SANMatchRules) == 0 {
+			log.Info().Msgf("mtls san match rules are undefined")
+			return nil
+		}
+		for _, match := range verifiedChains[0][0].DNSNames {
+			for _, candidateName := range s.SANMatchRules {
+				if matchExactly(match, candidateName) {
+					log.Debug().Msgf("Successfully matched matchExactly cert dns %s SANMatchRule %s", match, candidateName)
+					return nil
+				} else if matchHostnamesWithRegexp(match, candidateName) {
+					log.Debug().Msgf("Successfully matched matchHostnamesWithRegexp cert dns %s SANMatchRule %s", match, candidateName)
+					return nil
+				}
+			}
+		}
+
+		err = errors.New("certs verification with SAN match not found")
+		log.Error().Err(err).Msgf("SAN match rules %s", s.SANMatchRules)
+		return err
+	}
+}
+
+// toLowerCaseASCII returns a lower-case version of in. See RFC 6125 6.4.1. We use
+// an explicitly ASCII function to avoid any sharp corners resulting from
+// performing Unicode operations on DNS labels.
+func toLowerCaseASCII(in string) string {
+	// If the string is already lower-case then there's nothing to do.
+	isAlreadyLowerCase := true
+	for _, c := range in {
+		if c == utf8.RuneError {
+			// If we get a UTF-8 error then there might be
+			// upper-case ASCII bytes in the invalid sequence.
+			isAlreadyLowerCase = false
+			break
+		}
+		if 'A' <= c && c <= 'Z' {
+			isAlreadyLowerCase = false
+			break
+		}
+	}
+
+	if isAlreadyLowerCase {
+		return in
+	}
+
+	out := []byte(in)
+	for i, c := range out {
+		if 'A' <= c && c <= 'Z' {
+			out[i] += 'a' - 'A'
+		}
+	}
+	return string(out)
+}
+
+// matchExactly - match hostnames
+func matchExactly(hostA, hostB string) bool {
+	if hostA == "" || hostA == "." || hostB == "" || hostB == "." {
+		return false
+	}
+	return toLowerCaseASCII(hostA) == toLowerCaseASCII(hostB)
+}
+
+// matchHostnamesWithRegexp - To match the san rules with regexp
+func matchHostnamesWithRegexp(pattern, host string) bool {
```

Comment:
The parameters for this function are confusing.  Pattern makes me think it is the regex, but in fact it is the DNS Name from the Cert SANs.  Can we rename host to be something more obvious like MatchRegex